src/transformers/generation_tf_utils.py

```diff
@@ -677,21 +682,29 @@ def generate(
             # get encoder and store encoder outputs
             encoder = self.get_encoder()
 
-            encoder_outputs = encoder(
-                input_ids,
-                attention_mask=attention_mask,
-                output_attentions=output_attentions,
-                output_hidden_states=output_hidden_states,
-                return_dict=return_dict_in_generate,
-            )
+            encoder_kwargs = {
+                "attention_mask": attention_mask,
+                "output_attentions": output_attentions,
+                "output_hidden_states": output_hidden_states,
+                "return_dict": return_dict_in_generate,
+            }
+
+            # vision models don't use `attention_mask`.
+            signature = dict(inspect.signature(encoder.call).parameters)
+            if "attention_mask" not in signature:
+                encoder_kwargs.pop("attention_mask")
+
+            encoder_outputs = encoder(input_ids, **encoder_kwargs)
             if return_dict_in_generate:
                 if output_attentions:
                     model_kwargs["encoder_attentions"] = encoder_outputs.attentions
                 if output_hidden_states:
                     model_kwargs["encoder_hidden_states"] = encoder_outputs.hidden_states
 
+        # The condition `len(shape_list(input_ids)) == 2` is to make this block treats only text inputs.
+        # (vision inputs might occur when the model is an encoder-decoder model)
         # Expand input ids if num_beams > 1 or num_return_sequences > 1
-        if num_return_sequences > 1 or num_beams > 1:
+        if len(shape_list(input_ids)) == 2 and (num_return_sequences > 1 or num_beams > 1):
```

Comment:
that's a bit hacky - vision inputs should also work with `num_beams > 1` no? But ok for now until we do the big `generate` refactor @Rocketknight1

Reply:
Just as remark: The code inside this block treats `input_ids` as text-only inputs, for example `shape_list(input_ids)[-1]` assumes that the last dimension is the sequence dim.

For vision inputs, `generate` will be called only if it is a vision model as encoder in an encoder-decoder model (?). In this case, it is more the `decoder_input_ids` to be processed. And I think this is done in the next block

https://github.com/huggingface/transformers/blob/e2357599f7af1255b98729a3e0ef12bf2a081d44/src/transformers/generation_tf_utils.py#L722

It's not clear to me if an encoder-only vision model will need to call `generate()`. Maybe @NielsRogge can share some insights here (e.g. `ImageGPT`).